### PR TITLE
MCOL-4912-MCOL-4917 The patch replaces RBTree-based EM traversal with a

### PR DESCRIPTION
**Pull request by drrtuy**
combination of EM Index lookup + targeted RBTree search in EM routines
 This combination reduces avg search complexity from O(N) down to 0(1) * O(logN)
 that is crucial for EMs > 50MB.
 This patch also reduces RBTree-based EM footprint on shmem.

### CODE REVIEW

File: versioning/BRM/extentmap.cpp [hunk 1]
```diff
@@ -4034,23 +4001,18 @@ void ExtentMap::deleteOID(int OID)
...
-    while (it != end)
+    for (auto dbRoot: dbRootVec)
     {
-        if (it->second.fileID == OID)
+        const auto lbids = fPExtMapIndexImpl_->find(dbRoot, OID);
+        auto emIdents = getEmIteratorsByLbids(lbids);
+        OIDExists = (!emIdents.empty());
+        for (auto& emIt : emIdents)
         {
             OIDExists = true;
```
Comment (mariadb-DenisKhalikov) [inline versioning/BRM/extentmap.cpp]: Seems like this assignment is not needed anymore.

File: versioning/BRM/extentmap.cpp [hunk 2]
```diff
@@ -4924,30 +4871,19 @@ void ExtentMap::getExtentCount_dbroot(int OID, uint16_t dbroot, bool incOutOfSer
...
-        for (auto emIt = fExtentMapRBTree->begin(), end = fExtentMapRBTree->end(); emIt != end;
-             ++emIt)
+        const auto emIdents = getEmIdentsByLbids(lbids);
+        for (auto& emEntry : emIdents)
         {
-            const auto& emEntry = emIt->second;
-            if ((emEntry.fileID == OID) && (emEntry.dbRoot == dbroot) &&
-                (emEntry.status != EXTENTOUTOFSERVICE))
-                numExtents++;
+            numExtents = (emEntry.status != EXTENTOUTOFSERVICE) ? numExtents + 1 : numExtents;
```
Comment (mariadb-DenisKhalikov) [inline versioning/BRM/extentmap.cpp]: Nit: seems like this part will be complilled to something like this
```
if ()
  numExtents = numExtents + 1
else
  numExtents = numExtents
```
So, we will have an assignment on each branch, would it be better have only one branch and one assignment
```
if ()
  numExtents =  numExtents + 1
```
At least I'm not aware about ternary assembly command for x64

**Review by mariadb-DenisKhalikov — APPROVED** [review]
LGTM, I've added some NITS, but they are not blocking this PR.

**Review by mariadb-LeonidFedorov — APPROVED** [review]

File: versioning/BRM/extentmap.cpp (hunk 1 above)
Comment (drrtuy) [reply versioning/BRM/extentmap.cpp]: Done.

File: versioning/BRM/extentmap.cpp (hunk 2 above)
Comment (drrtuy) [reply versioning/BRM/extentmap.cpp]: Agree.
Fixed.

**Pull request merged**